compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUMaterializeEncodingPass.cpp

```diff
@@ -0,0 +1,118 @@
+#include "iree-dialects/Dialect/LinalgExt/Passes/Passes.h"
+#include "iree/compiler/Codegen/Common/EncodingInfo.h"
+#include "iree/compiler/Codegen/PassDetail.h"
+#include "iree/compiler/Codegen/Utils/Utils.h"
+#include "iree/compiler/Dialect/Flow/IR/FlowOps.h"
+#include "iree/compiler/Dialect/HAL/IR/HALTypes.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/MemRef/Transforms/Passes.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir {
+namespace iree_compiler {
+
+using namespace IREE::LinalgExt;
+using IREE::HAL::ExecutableTargetAttr;
+
+namespace {
+
+static MatmulTileParams chooseMatmulTileParamsGeneric() { return {8, 4, 8}; }
+
+static MatmulTileParams chooseMatmulTileParamsAArch64(
+    MatmulType type, ExecutableTargetAttr target) {
+  switch (type) {
+    case MatmulType::F32F32F32:
+      return {8, 1, 8};
+    case MatmulType::I8I8I32:
+      if (hasFeature(target, "+i8mm")) return {8, 8, 8};
+      if (hasFeature(target, "+dotprod")) return {8, 4, 8};
+      return {8, 1, 8};
+    default:
+      assert(false);
+      return {};
+  }
+}
+
+static MatmulTileParams chooseMatmulTileParams(MatmulType type,
+                                               ExecutableTargetAttr target) {
+  if (isAArch64(target)) {
+    return chooseMatmulTileParamsAArch64(type, target);
+  }
+  return chooseMatmulTileParamsGeneric();
+}
+
+struct LLVMCPUMaterializeEncodingPass
+    : public LLVMCPUMaterializeEncodingBase<LLVMCPUMaterializeEncodingPass> {
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<arith::ArithDialect, AffineDialect, IREE::Flow::FlowDialect,
+                    IREE::LinalgExt::IREELinalgExtDialect>();
+  }
+  void runOnOperation() override;
+};
+
+}  // namespace
+
+void LLVMCPUMaterializeEncodingPass::runOnOperation() {
+  MLIRContext *context = &getContext();
+  auto operation = getOperation();
+  RewritePatternSet materializeEncodingPattern(context);
+  auto targetAttr = ExecutableTargetAttr::lookup(operation);
+  MaterializeEncodingTypeConverter typeConverter(
+      [targetAttr](
+          RankedTensorType tensorType) -> FailureOr<MaterializeEncodingInfo> {
+        Optional<TensorEncoding> encoding = getEncoding(tensorType);
+        if (!encoding) return failure();
+
+        auto matmulType = getMatmulType(*encoding);
+        auto matmulOperandRole = getMatmulOperandRole(*encoding);
+        if (!matmulType || !matmulOperandRole) {
+          return failure();
+        }
+        MatmulTileParams tileParams =
+            chooseMatmulTileParams(*matmulType, targetAttr);
+        auto encodingInfo = chooseEncodingInfoForMatmul(
+            *matmulType, *matmulOperandRole, tileParams);
+        adjustTileSizesToNarrowStaticShape(encodingInfo, tensorType.getShape());
+        return encodingInfo;
+      });
```

Comment:
Not really sure how we can refactor this out. LLVMCPU and VMVX basically have the same implementation except calling `chooseMatmulTileParams` method.

Reply:
It is OK to duplicate for now.... If all backends were going to do the same thing then its worth having a common pass, but as we discussed there might be backend-specific things here that might be needed.